File: acruiseR/R/peakid.R
#' Identifies background from a concentration time-series.
#'
#' This process takes 3 steps:
#' 1. Obtaining a smoothed rolling standard deviation of the background
#' 2. Identifying background measurements as those that lie within a set
#' threshold
#' 3. Interpolating values outside of this background threshold in a
#' linear manner and then using a rolling mean so that 'background'
#' measurements are available for the entire time-series.
#'
#' @param concentration Concentration time-series as a vector.
#' @param method The background identification method to be employed.
#' \itemize{
#'     \item{\code{gam}: Fits a Generalised Additive Model (gam), which is a non-linear
#'     regression technique. Requires parameter \code{k} to be tuned.}
#'     \item{\code{rolling}: The original 2-step rolling average method, which
#'     firstly fits a rolling standard deviation and then a rolling mean to extract
#'     the background. Requires parameters \code{bg_sd_window}, \code{bg_sd_threshold},
#'     and \code{bg_mean_window} to be tuned.}
#' }
#' @param k The number of spline points in the GAM, a higher number
#'       should provide a better fit but at the cost of overfitting to
#'       noise. Used only
#'       when \code{method='gam'}.
#' @param bg_sd_window Window size for the rolling standard deviation
#'       smooth to identify the background, as an integer. Used only
#'       when \code{method='rolling'}.
#' @param bg_sd_threshold Background measurements are considered as
#'       those whose rolling sd is within this threshold
#'       when \code{method='rolling'}. Used only
#' @param bg_mean_window The rolling mean to smooth the interpolated
#'       background, as an integer. Used only
#'       when \code{method='rolling'}.
#'
#' @return A list containing the background (slot `bg`) which is
#' a vector with the same length as `concentration`, and the arguments
#' used in this call (slot `call`).
#' @export
identify_background <- function(concentration,
                                method = c("gam", "rolling"),
                                k = 10,
                                bg_sd_window = 180,
                                bg_sd_threshold = 0.5,
                                bg_mean_window = 660) {
    method <- match.arg(method)
    if (method == "rolling") {
        # Smooth concentration to identify background values (low SD threshold)
        bg <- RcppRoll::roll_sd(data.table::nafill(concentration, type = "locf"),
            n = bg_sd_window,
            align = "center",
            fill = NA
        )
        # With even windows there is a non-even number of NAs around the output
        # Pandas the extra NA to the start, R places it at the end
        # This line make R consistent with pandas
        if (bg_sd_window %% 2 == 0) {
            bg <- c(NA, bg[1:length(bg) - 1])
        }
        # Average
        bg <- RcppRoll::roll_mean(bg, n = bg_sd_window, align = "center", fill = NA)
        # Again move surplus NA to start
        if (bg_sd_window %% 2 == 0) {
            bg <- c(NA, bg[1:length(bg) - 1])
        }

        is_bg <- !is.na(concentration) & !is.na(bg) & bg <= bg_sd_threshold

        # Now take the original raw concentration and linearly interpolate any non-background
        # values, and then take a final smooth
        output <- rep(NA, length(concentration))
        output[is_bg] <- concentration[is_bg]
        # If want to interpolate everything, so including the NAs introduced by the
        # rolling functions then use rule=2
        output[!is_bg] <- approx(output, xout = which(!is_bg), rule = 1)$y
        bg_out <- RcppRoll::roll_mean(output, n = bg_mean_window, align = "right", fill = NA)
    } else if (method == "gam") {
        # Interpolate missing values, unlike the first method that uses LastOneCarryForward
        concentration <- forecast::na.interp(concentration)
        x <- seq_along(concentration)

        # Fit baseline
        mod <- mgcv::gam(concentration ~ s(x, bs = "cs", k = k), method = "REML", select = TRUE)
        bg_out <- mod$fitted.values
    }
    out <- list(
        bg = bg_out,
        call = list(
            method = method, k = k, bg_sd_window = bg_sd_window,
            bg_sd_threshold = bg_sd_threshold, bg_mean_window = bg_mean_window
        )
    )
    out
}

#' Detects plumes in a concentration time series.
#'
#' @inheritParams identify_background
#' @param background The smoothed background time-series object, as can be
#' obtained from `identify_background`
#' @param time The time-series as a vector of POSIX or nanotime.
#' Must have the same length as `concentration`.
#' @param plume_sd_threshold The number of standard deviations that a sample
#' must exceed to be defined as a plume.
#' @param plume_sd_starting Once a plume has been identified due to it
#' crossing \code{plume_sd_threshold}, its duration is considered as the
#' times when it is more than this many standard deviations above the
#' background.
#' @param plume_buffer A buffer in seconds applied to plumes, so
#' that if they are overlapping they are merged into the same plume.
#' @param refit If TRUE, removes the identified plumes from the time-series
#' to recalculate the background, then re-identifies plumes. Should result in
#' a smoother background and therefore more accurate plume estimation.
#'
#' @return A Data Frame where each row corresponds to a unique plume, whose
#' time boundaries are contained in the 2 columns: `start` and `end`.
#' @import data.table
#' @export
detect_plumes <- function(concentration,
                          background,
                          time,
                          plume_sd_threshold = 3,
                          plume_sd_starting = 2,
                          plume_buffer = 10,
                          refit = FALSE) {
    # Convert into nanotime (64-bit int) if in POSIX (32-bit double)
    time <- posix_to_nanotime(time)

    # First pass at identifying plumes
    bg <- background$bg
    residual_sd <- sd(concentration - bg, na.rm = T)
    dt <- data.table::data.table(time = time, concentration = concentration, background = bg)
    setorder(dt, time)
    dt[, is_plume_starting := !is.na(concentration) & !is.na(background) & concentration > (background + plume_sd_starting * residual_sd)]
    dt[, is_plume := !is.na(concentration) & !is.na(background) & concentration > (background + plume_sd_threshold * residual_sd)]
    dt[, plume_group_starting := cumsum((is_plume_starting != data.table::shift(is_plume_starting, fill = F, type = "lag")))]

    plume_groups_dt <- dt[is_plume_starting == TRUE, list("has_plume" = sum(is_plume), start = min(time), end = max(time)), by = plume_group_starting][has_plume > 0]
    if (nrow(plume_groups_dt) == 0) {
        return(data.table(start=nanotime(), end=nanotime()))
    }

    # Find overlapping plumes within the buffer period
    setkey(plume_groups_dt, start, end)
    overlaps <- foverlaps(plume_groups_dt,
        plume_groups_dt[, .(start, end = end + nanotime::nanoduration(hours = 0, minutes = 0, seconds = plume_buffer, nanoseconds = 0))],
        type = "any",
        which = TRUE
    )
    overlaps[, c("x_prev_seen", "y_prev_seen") := list(duplicated(xid), duplicated(yid))]
    overlaps[, combined_plume := cumsum(!(x_prev_seen | y_prev_seen))]
    # Find group for each row number
    new_plumes <- unique(overlaps, by = c("xid", "combined_plume"))
    # Join back into overlapping plumes then elongate plumes with new members
    plumes_final <- plume_groups_dt[, xid := 1:nrow(plume_groups_dt)][new_plumes, on = "xid"][, list(
        start = min(start),
        end = max(end)
    ),
    by = combined_plume
    ]

    # Format for output
    plumes_final[, combined_plume := NULL]
    setorder(plumes_final, start)
    out <- as.data.frame(plumes_final)

    if (refit) {
        # Remove plumes and reidentify background
        plumes_final[, in_plume := TRUE]
        plumes_removed <- plumes_final[dt, on = c("start <= time", "end >= time"), .(time, concentration, background, in_plume)]
        plumes_removed[in_plume == TRUE, concentration := NA]
        args <- c(list(concentration = plumes_removed$concentration), background$call)
        bg_new <- do.call(identify_background, args)

        # Redetect plumes with new background
        out <- detect_plumes(concentration,
            bg_new,
            time,
            plume_sd_threshold = plume_sd_threshold,
            plume_sd_starting = plume_sd_starting,
            plume_buffer = plume_buffer,
            refit = FALSE
        )
    }
    out
}

#' Integrate the Area Under a Plume (aup) using a trapezoidal method.
#'
#' @inheritParams detect_plumes
#' @param plumes A Data Frame with 'start' and 'end' columns
#' containing plume boundaries, as returned by `detect_plumes`.
#' @param dx Sampling period, passed onto the dz argument of
#' np.trapz. I.e. the time between consecutive measurements.
#' @param uncertainty Instrument uncertainty as a decimal (i.e. 5% is 0.05).
#' If provided then corresponding uncertainty estimates are also provided, if
#' `NULL` then these are omitted.
#'
#' @return A Data Frame with one row per plume and 3 columns `start`, `end`, and
#' `area`, and an additional `uncertainty` column if the `uncertainty` parameter is
#' not `NULL`. The first 2 columns are the same as in the input `plumes`, while `area`
#' contains the integrated area.
#' @import data.table
#' @export
integrate_aup_trapz <- function(concentration, time, plumes, dx = 1, uncertainty = NULL,
                                uncertainty_type = c("absolute", "relative")) {
    uncertainty_type <- match.arg(uncertainty_type)
    # Ensure both time columns are in the same format. Could stick with both in POSIX
    # but might as we use nanotime
    time <- posix_to_nanotime(time)
    plumes$start <- posix_to_nanotime(plumes$start)
    plumes$end <- posix_to_nanotime(plumes$end)
    # Join plumes into the main concentration so can calculate area by plume
    dt <- data.table(concentration = concentration, time = time)
    plumes_dt <- data.table(plumes)
    plumes_dt[, plume_id := 1:nrow(plumes_dt)]

    areas <- dt[plumes_dt, on = c("time >= start", "time <= end")][, .(area = pracma::trapz(seq(length.out = .N, by = dx), concentration), concentration), by = list(time, time.1, plume_id)]

    if (!is.null(uncertainty)) {
        h2_2 <- (dx / 2)**2
        if (uncertainty_type == "relative") {
            areas[, delta := h2_2 * (uncertainty * concentration)**2]
        } else if (uncertainty_type == "absolute") {
            areas[, delta := h2_2 * (uncertainty)**2]
        }
        # For each consecutive point within a plume, take the Sqrt of the sum of each point's uncertainty
        areas <- areas[, .(area = mean(area, na.rm = T), uncertainty = sum(sqrt(delta + lag(delta)), na.rm = T)), by = .(time, time.1, plume_id)]
    }

    areas[, plume_id := NULL]
    setnames(areas, c("time", "time.1"), c("start", "end"))
    setcolorder(areas, c("start", "end", "area"))
    areas
}

#' Diagnostic plot to aid background extraction
#'
#' Plots the concentration time-series highlighting the extracted background
#' (red), the threshold for what is considered a plume (orange), and when
#' plumes will be determined to have started at (blue).
#'
#' @inheritParams detect_plumes
#' @param ylabel y-axis label
#' @param xlabel x-axis label
#' @param date_fmt Format of the x-axis labels, see \code{date_labels}
#' argument of \code{ggplot2::scale_x_datetime}.
#' @param bg_alpha The alpha level of the background concentration.
#' @return A \code{ggplot2} object, so it can be modified further.
#' @export
plot_background <- function(concentration,
                            time,
                            background,
                            plume_sd_threshold = 3,
                            plume_sd_starting = 2,
                            ylabel = "Concentration",
                            xlabel = "Time (UTC)",
                            date_fmt = "%H:%M",
                            bg_alpha = 0.5) {
    .Deprecated("plot_plumes")
    background <- background$bg
    time <- nanotime_to_posix(time) # Can't plot nanotime
    dt <- data.table(concentration = concentration, time = time, bg = background)
    sd_residual <- sd(concentration - background, na.rm = T)
    dt[, bg_starting := bg + plume_sd_starting * sd_residual]
    dt[, bg_threshold := bg + plume_sd_threshold * sd_residual]
    dt <- melt(dt, id.vars = "time")
    dt[, variable := factor(variable,
        levels = c("concentration", "bg", "bg_starting", "bg_threshold"),
        labels = c("Concentration", "Mean background", "Plume starting point", "Plume threshold")
    )]

    ggplot2::ggplot(dt, ggplot2::aes(x = time, y = value, colour = variable, alpha = variable)) +
        ggplot2::geom_line(na.rm=TRUE) +
        ggplot2::labs(x = xlabel, y = ylabel) +
        ggplot2::scale_x_datetime(date_labels = date_fmt) +
        ggplot2::scale_colour_manual("", values = c("gray", "red", "steelblue", "orange")) +
        ggplot2::scale_alpha_manual("", values = c(bg_alpha, 1, 1, 1)) +
        ggplot2::theme_minimal()
}

#' Plots plumes against the background concentration.
#'
#' @inheritParams integrate_aup_trapz
#' @inheritParams plot_background
#'
#' @return A \code{ggplot2} object, so it can be modified further.
#' @export
plot_plumes <- function(concentration,
                        time,
                        plumes=NULL,
                        background=NULL,
                        ylabel = "Concentration",
                        xlabel = "Time (UTC)",
                        date_fmt = "%H:%M",
                        plume_sd_threshold = 3,
                        plume_sd_starting = 2,
                        bg_alpha = 0.5) {
    dt <- data.table(concentration = concentration, time = time)

    # Straight away remove missing valuesjk
    dt <- dt[ !is.na(concentration)]

    # Create empty plumes if not provided, just makes plotting easier
    if (is.null(plumes)) {
        plumes <- data.table(start = nanotime(), end = nanotime())
    }
    plumes_dt <- as.data.table(plumes)
    plumes_dt[, plume_id := 1:nrow(plumes_dt)]
    dt <- plumes_dt[dt, on = c("start <= time", "end >= time"),
                    .(time, plume_id, concentration)]
    setorder(dt, time)  # Shouldn't be needed but best to do
    dt[, c('plume_id', 'is_plume') := .(as.factor(plume_id),
                                        factor(is.na(plume_id),
                                               levels=c(FALSE, TRUE)))]

    if (!is.null(background)) {
        bg <- background$bg
        sd_residual <- sd(concentration - bg, na.rm = T)
        dt[, c('bg', 'bg_starting', 'bg_threshold') := list(bg,
                                                         bg + plume_sd_starting * sd_residual,
                                                         bg + plume_sd_threshold * sd_residual
        )]
    }

    # Manually repeat high contrast colour palette to match number of plumes,
    # as palette only has 9 values. This mimics matplotlib's behaviour
    n_plumes <- nrow(plumes_dt)
    max_set_1 <- 7 # The 8th (last) colour in the palette is grey, which clashes with background. Remove it.
    palette <- "Dark2"
    if (n_plumes == 0) {
        colours <- c('grey50')
        alphas <- bg_alpha
    } else if (n_plumes < 3) {  # Dark2 needs 3 colours minimum
        colours <- RColorBrewer::brewer.pal(3, palette)
        colours <- colours[1:n_plumes]
        alphas <- c(1, bg_alpha)
    } else if (n_plumes <= max_set_1) {
        colours <- RColorBrewer::brewer.pal(n_plumes, palette)
        alphas <- c(1, bg_alpha)
    } else {
        n_repeats <- floor(n_plumes / max_set_1)
        mod <- max(n_plumes %% max_set_1, 3) # Can't request < 3 colours
        colours <- c(
            rep(RColorBrewer::brewer.pal(max_set_1, palette), times = n_repeats),
            RColorBrewer::brewer.pal(mod, palette)
        )
        alphas <- c(1, bg_alpha)
    }

    dt[, time := nanotime_to_posix(time)] # Can't plot nanotime
    # Only want to plot non-missing value as otherwise ggplot can disappear lines
    dt <- dt[ !is.na(concentration)]

    p <- ggplot2::ggplot(dt, ggplot2::aes(x = time,
                                          y = concentration)) +
        ggplot2::geom_line(ggplot2::aes(colour = plume_id, alpha=is_plume), na.rm=TRUE) +
        ggplot2::labs(x = xlabel, y = ylabel) +
        ggplot2::scale_x_datetime("", date_labels = date_fmt, timezone = tz(dt$time)) +
        ggplot2::guides(colour = "none", alpha = "none") +
        ggplot2::scale_colour_manual("", values = colours) +
        ggplot2::scale_alpha_manual("", values = alphas) +
        ggplot2::theme_minimal()

    if (!is.null(background)) {
        p <- p + ggplot2::geom_line(aes(y=bg), colour="red", na.rm=TRUE) +
                 ggplot2::geom_line(aes(y=bg_starting), colour="steelblue", na.rm=TRUE) +
                 ggplot2::geom_line(aes(y=bg_threshold), colour="orange", na.rm=TRUE)
    }
    p
}

posix_to_nanotime <- function(x) {
    if (any(grepl("POSIX", class(x)))) {
        tz <- attr(x, "tz")
        if (tz == "") {
            tz <- "UTC"
        }
        x <- nanotime::as.nanotime(strftime(x, "%Y-%m-%d %H:%M:%OS3"), format = "%Y-%m-%d %H:%M:%ES", tz = tz)
    }
    x
}

nanotime_to_posix <- function(x) {
    if (any(grepl("nanotime", class(x)))) {
        tz <- "UTC"
        x <- as.POSIXct(x, tz = tz)
    }
    x
}

File: acruisepy/acruisepy/peakid.py
from functools import reduce
import datetime
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
import numpy as np
import matplotlib.dates as mdates
import pywt
from typing import Optional
from warnings import warn


def identify_background(
    concentration: pd.Series,
    bg_sd_window: int = 180,
    bg_sd_threshold: float = 0.5,
    bg_mean_window: int = 660,
) -> pd.Series:
    """
    Identifies background from a concentration time-series.

    This process takes 3 steps:

        - Obtaining a smoothed rolling standard deviation of the background
        - Identifying background measurements as those that lie within a set
          sd threshold
        - Interpolating values outside of this background threshold in a
          linear manner and then using a rolling mean so that 'background'
          measurements are available for the entire time-series.

    Args:
        - concentration (pd.Series): Concentration time-series.
        - bg_sd_window (int): Window size for the rolling standard deviation
          smooth to identify the background.
        - bg_sd_threshold (float): Background measurements are considered as
          those whose rolling sd is within this threshold.
        - bg_mean_window (int): The rolling mean to smooth the interpolated
          background

    Returns:
        The smoothed background covering the full times series as a pd.Series
        object.
    """
    warn('The rolling window method is deprecated in favour of the wavelet detection. Please refer to the README.', DeprecationWarning, stacklevel=2)
    # Smooth concentration to get smoothed rolling SD
    roll_std = (
        concentration.fillna(method="ffill")
        .rolling(bg_sd_window, center=True)
        .std()
        .rolling(bg_sd_window, center=True)
        .mean()
    )
    # Define background as when rolling std deviation is below a threshold
    is_bg = (~concentration.isna()) & (roll_std <= bg_sd_threshold)
    # Interpolate background values when don't have background
    # If want to interpolate everything, so including the NAs introduced by the
    # rolling functions then use limit_direction="both"
    bg = (
        concentration.loc[is_bg]
        .reindex(concentration.index)
        .interpolate(limit_area="inside")
        .rolling(bg_mean_window)
        .mean()
    )

    return bg


def detect_plumes(
    concentration: pd.Series,
    background: pd.Series,
    plume_sd_threshold: float = 3,
    plume_sd_starting: float = 2,
    plume_buffer: float = 10,
) -> pd.DataFrame:
    """
    Detects plumes in a concentration time series.

    Args:
        - concentration (pd.Series): The concentration time-series. Must have a
            Datetime index.
        - background (pd.Series): The smoothed background time-series, as can be
            obtained from identify_background(). Must have the same
            Datetime index as concentration.
        - plume_sd_threshold (float): Plumes are identified as samples
            greater than certain number of standard deviations from a
            smoothed background.
        - plume_sd_starting (float): For any identified plumes, it is determined
            to have started once it reached this many standard deviations above
            the background.
        - plume_buffer (float): A buffer in seconds applied to plumes, so
            that if they are overlapping they are merged into the same plume.

    Returns:
        A pd.DataFrame where each row corresponds to a unique plume, whose
        time boundaries are contained in the 2 columns: `start` and `end`.
    """
    warn('The rolling window method is deprecated in favour of the wavelet detection. Please refer to the README.', DeprecationWarning, stacklevel=2)
    df = pd.DataFrame({"concentration": concentration, "background": background})
    # Rename index so can reliably refer to it later
    df.index.rename("index", inplace=True)

    # Derive useful values for identifying plumes
    df["is_plume"] = df["concentration"] > (
        df["background"] + plume_sd_threshold * df["background"].std()
    )
    df["is_plume_starting"] = df["concentration"] > (
        df["background"] + plume_sd_starting * df["background"].std()
    )
    df["plume_group_starting"] = (
        df["is_plume_starting"] != df["is_plume_starting"].shift()
    ).cumsum()

    # Find all groups where concentration are > than the lower starting threshold, and that
    # also have at least one value greater than the higher threshold required to say it is a plume
    plume_groups = (
        df.loc[df["is_plume_starting"]]
        .reset_index()
        .groupby("plume_group_starting")
        .agg(
            has_plume=pd.NamedAgg(column="is_plume", aggfunc="sum"),
            start=pd.NamedAgg(column="index", aggfunc="min"),
            end=pd.NamedAgg(column="index", aggfunc="max"),
        )
        .query("has_plume > 0")
        .drop("has_plume", axis=1)
        .sort_values(["start", "end"])
    )

    # Combine overlapping plumes
    buffer_time = datetime.timedelta(seconds=plume_buffer)
    # Use Interval data structure here which is effectively syntatical sugar
    # around a tuple
    plume_intervals = [
        pd.Interval(s, e, closed="both")
        for s, e in zip(plume_groups["start"], plume_groups["end"])
    ]

    def reduce_intervals(acc, el):
        plume_with_buffer = pd.Interval(acc[0].left, acc[0].right + buffer_time)
        if plume_with_buffer.overlaps(el):
            return [pd.Interval(acc[0].left, el.right)] + acc[1:]
        else:
            return [el] + acc

    plume_overlap = reduce(reduce_intervals, plume_intervals[1::], [plume_intervals[0]])

    # Convert into DataFrame for user friendliness
    plumes_condensed = pd.DataFrame(
        [{"start": x.left, "end": x.right} for x in plume_overlap]
    ).sort_values(["start"])

    output = []
    plumes_condensed['plume_id'] = range(1, plumes_condensed.shape[0]+1)
    for row in plumes_condensed.itertuples():
        sub_df = df.loc[row.start : row.end].copy(deep=True)
        sub_df['plume_id'] = row.plume_id
        output.append(sub_df)
    output = pd.concat(output)
    return output[['plume_id', 'concentration']]


def integrate_aup_trapz(
    concentration: pd.Series,
    plumes: pd.DataFrame,
    background: Optional[pd.Series] = None,
    dx: Optional[float] = 1.0
) -> pd.DataFrame:
    """
    Integrate the Area Under a Plume (aup) using a trapezoidal method.

    Args:
        - concentration (pd.Series): The concentration time-series. 
          Must have a Datetime index.
        - plumes (pd.DataFrame): A DataFrame as returned by `detect_plumes()`.
        - background (pd.Series): A time-series of background measurements, the
          same length as `concentration`. Used to subtract the background before
          peak integration. If not provided interpolates linearly over the plume
          duration. Must share an index with `concentration`.
        - dx (float): Sampling time, passed onto the dx argument of
          np.trapezoid.

    Returns:
        A pd.DataFrame with one row per plume and 4 columns:
            - `plume_id`: Integer plume ID, as obtained in `detect_plumes_wavelets`
            - `start`: Start time of this plume
            - `end`: End time of this plume
            - `area`: Total area under the plume.
    """
    if background is None:
        background = concentration.copy(deep=True)
        for plume_id, sub_df in plumes.groupby('plume_id'):
            background.loc[sub_df.index.min():sub_df.index.max()] = np.nan
        background = background.interpolate(method='linear')
    concentration = concentration - background

    areas = []
    for plume_id, sub_df in plumes.groupby('plume_id'):
        start = sub_df.index.min()
        end = sub_df.index.max()
        this_df = pd.DataFrame(
            [
                {
                    "plume_id": plume_id,
                    "start": start,
                    "end": end,
                    "area": np.trapezoid(concentration.loc[start:end], dx=dx),
                }
            ]
        )
        areas.append(this_df)
    return pd.concat(areas)


def plot_background(
    concentration: pd.Series,
    background: pd.DataFrame,
    plume_sd_threshold: float = 3,
    plume_sd_starting: float = 2,
    ylabel: str = "Concentration",
    xlabel: str = "Time (UTC)",
    date_fmt: str = "%H:%M",
    bg_alpha: float = 0.5,
) -> None:
    """
    Plots the concentration time-series highlighting the extracted background
    (red), the threshold for what is considered a plume (orange), and when
    plumes will be determined to have started at (blue).

    Args:
        - concentration (pd.Series): The concentration time-series. Must have a Datetime
            index.
        - background (pd.Series): The background time-series, as obtained
          from identify_background(). Must have a Datetime index.
        - plume_sd_threshold (float): Plumes are identified as samples
            greater than certain number of standard deviations from a
            smoothed background.
        - plume_sd_starting (float): For any identified plumes, it is determined
            to have started once it reached this many standard deviations above
            the background.
        - ylabel (str): y-axis label
        - xlabel (str): x-axis label
        - date_fmt (str): How to display the x-axis datetime breaks
        - bg_alpha (float): Alpha level of the background concentration.

    Returns:
        None, plots a figure as a side-effect.
    """
    warn('The rolling window method is deprecated in favour of the wavelet detection. Please refer to the README.', DeprecationWarning, stacklevel=2)
    fig, ax = plt.subplots()
    myFmt = mdates.DateFormatter(date_fmt)
    ax.xaxis.set_major_formatter(myFmt)
    ax.set_ylabel(ylabel)
    ax.set_xlabel(xlabel)
    ax.plot(concentration, color="gray", alpha=bg_alpha)
    ax.plot(background, color="red", label="Mean background")
    ax.plot(
        background + plume_sd_threshold * background.std(),
        color="orange",
        label="Plume threshold",
    )
    ax.plot(
        background + plume_sd_starting * background.std(),
        color="steelblue",
        label="Plume starting point",
    )
    ax.legend()
    plt.show()


def plot_plumes(
    concentration: pd.Series,
    plumes: pd.DataFrame,
    ylabel: str = "Concentration",
    xlabel: str = "Time (UTC)",
    date_fmt: str = "%H:%M",
    bg_alpha: float = 0.5,
) -> None:
    """
    Plots plumes against the background concentration.

    Args:
        - concentration (pd.Series): The concentration time-series with the background
          removed. Must have a Datetime index.
        - plumes (pd.DataFrame): A DataFrame with 'start' and 'end' columns
          containing plume boundaries, as returned by detect_plumes()
        - ylabel (str): y-axis label
        - xlabel (str): x-axis label
        - date_fmt (str): How to display the x-axis datetime breaks
        - bg_alpha (float): Alpha level of the background concentration.

    Returns:
        None, plots a figure as a side-effect.
    """
    fig, ax = plt.subplots()
    myFmt = mdates.DateFormatter(date_fmt)
    ax.xaxis.set_major_formatter(myFmt)
    ax.set_ylabel(ylabel)
    ax.set_xlabel(xlabel)
    ax.plot(concentration, color="gray", alpha=bg_alpha)
    for plume_id, sub_df in plumes.groupby('plume_id'):
        ax.plot(sub_df['concentration'])
    plt.show()

def max_wavelet_level(concentration: pd.Series):
    """
    Identifies the maximum number of levels a signal can be decomposed into
    using wavelets.

    This is a thin wrapper around `pywt.dwt_max_level`.

    Args:
        concentration (pd.Series): The input concentration time-series.

    Returns:
        An integer with the maximum number of levels that can be used in
        `detect_plumes_wavelets`.
    """
    return pywt.dwt_max_level(len(concentration), 'haar')

def detect_plumes_wavelets(concentration: pd.Series,
                           levels: Optional[list[int]] = None,
                           plume_threshold: float =1,
                           plume_starting: float =0.5,
                           plume_buffer: float = 10,
                           interpolate: bool = False,
                           plot: bool = False):
    """
    Detects peaks using a wavelet decomposition.

    This function works differently to the main `detect_plumes`. Rather than
    taking a 2-step approach to identify the background and then remove it leaving
    the plumes behind, the plumes are instead directly identified.

    The concentration time-series is decomposed using a multi-level Haar
    wavelet.
    By reconstructing the signal using a partial set of lower-order Wavelet
    coefficients, the resulting signal has the high-frequency plume content but
    without the slowly-varying background.
    This is a very quick process, so it is recommended to iteratively try
    different levels and plotting the results (with `plot=True`) before
    saving the output plume locations.

    For a gentle introduction to the Haar wavelet, see the first 11 minutes of
    this video:
    https://www.youtube.com/watch?v=c1XL5BeI9_s

    Args:
        - concentration (pd.Series): Concentration time-series, must have a 
            Datetime index.
        - levels (list[int]): Which levels of the decomposition to use when
            reconstructing the peaks
        - plume_threshold (float): The threshold determine whether a signal is a
            plume.
        - plume_starting (float): The threshold from where plumes are determined
            to start.
        - plume_buffer (float): A buffer in seconds applied to plumes, so
            that if they are overlapping they are merged into the same plume.
        - interpolate (bool): Whether to linearly interpolate missing values
            from the raw concentration time-series. Can help in certain conditions.
        - plot (bool): Whether to plot a diagnostic plot of the wavelet
            recomposition with the threshold lines. Used to help identify optimal
            parameter settings.

    Returns:
        A pd.DataFrame where each row corresponds to a timepoint within a plume
        with 3 columns:
          - plume_id: Integer ID identifying the plume this measurement came from
          - concentration: Raw concentration at this timepoint
          - reconstruction: Concentration with the background removed
    """
    # Interpolate missing values - can mess with Wavelets
    if interpolate:
        concentration = concentration.interpolate()
    coefs = pywt.wavedec(concentration, 'haar')
    max_level = max_wavelet_level(concentration)

    # TODO Just pass number levels and let it work out levels for us
    # Wrap levels in list if provided single int
    if type(levels) is int:
        levels = [levels]
    if any(max_level < l < 1 for l in levels):
        raise ValueError("Levels must be between 1 and a maximum level, determined by `max_wavelet_level`")

    # Set unselected levels to zero
    # NB since only ever want the detail and not the approximation (index 0).
    # This has the handy bonus effect of meaning we don't need to explicitly
    # convert between 1-index (User-interface) and 0-index (how they are stored)
    if levels is not None:
        coefs_selected = [np.zeros_like(x) if i not in levels else x for i, x in enumerate(coefs)]

    recon = abs(pywt.waverec(coefs_selected, 'haar'))

    if plot:
        fig, ax = plt.subplots()
        ax.plot((concentration - concentration.mean()).reset_index(drop=True), label=f"Normalised raw signal", alpha=0.5)
        ax.plot(recon, label=f"Reconstructed signal using levels {','.join((str(x) for x in levels))}")
        ax.hlines(plume_threshold, xmin=0, xmax=len(recon), colors='C2',
                   label="plume_threshold")
        ax.hlines(plume_starting, xmin=0, xmax=len(recon), colors='C3',
                   label="plume_starting")
        ax.legend()
        plt.show()

    # Ensure the 2 signals are the same length - not guaranteed!
    con_start = np.max(np.array([0, concentration.size-recon.size]))
    recon_end = recon.size if recon.size <= concentration.size else concentration.size - recon.size
    df = pd.DataFrame({"concentration": concentration[con_start:], "reconstruction": recon[:recon_end]})
    # Rename index so can reliably refer to it later
    df.index.rename("index", inplace=True)

    # Derive useful values for identifying plumes
    df["is_plume"] = df["reconstruction"] > plume_threshold
    df["is_plume_starting"] = df["reconstruction"] > plume_starting
    df["plume_group_starting"] = (
        df["is_plume_starting"] != df["is_plume_starting"].shift()
    ).cumsum()

    # Find all groups where concentration are > than the lower starting threshold, and that
    # also have at least one value greater than the higher threshold required to say it is a plume
    plume_groups = (
        df.loc[df["is_plume_starting"]]
        .reset_index()
        .groupby("plume_group_starting")
        .agg(
            has_plume=pd.NamedAgg(column="is_plume", aggfunc="sum"),
            start=pd.NamedAgg(column="index", aggfunc="min"),
            end=pd.NamedAgg(column="index", aggfunc="max"),
        )
        .query("has_plume > 0")
        .drop("has_plume", axis=1)
        .sort_values(["start", "end"])
    )

    # Combine overlapping plumes
    buffer_time = datetime.timedelta(seconds=plume_buffer)
    # Use Interval data structure here which is effectively syntatical sugar
    # around a tuple
    plume_intervals = [
        pd.Interval(s, e, closed="both")
        for s, e in zip(plume_groups["start"], plume_groups["end"])
    ]

    def reduce_intervals(acc, el):
        plume_with_buffer = pd.Interval(acc[0].left, acc[0].right + buffer_time)
        if plume_with_buffer.overlaps(el):
            return [pd.Interval(acc[0].left, el.right)] + acc[1:]
        else:
            return [el] + acc

    plume_overlap = reduce(reduce_intervals, plume_intervals[1::], [plume_intervals[0]])

    # Convert into DataFrame for user friendliness
    plumes_condensed = pd.DataFrame(
        [{"start": x.left, "end": x.right} for x in plume_overlap]
    ).sort_values(["start"])

    # Join back into time-series and extract concentration
    output = []
    plumes_condensed['plume_id'] = range(1, plumes_condensed.shape[0]+1)
    for row in plumes_condensed.itertuples():
        sub_df = df.loc[row.start : row.end].copy(deep=True)
        sub_df['plume_id'] = row.plume_id
        output.append(sub_df)
    output = pd.concat(output)

    return output[['plume_id', 'concentration']]
